### DISCUSSION TITLE
Custom syntax blocks

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [vlang/v#11767] [Ideas]
Whilst browsing `vlib/v` and taking a good look at the tokenizer, scanner, parser and generators, I noticed that there are a few special cases: `$vweb.html()`, `db mydb {...}` and the already documented `asm x86 {...}`. Especially SQL and VWeb are quite specific. But, what if someone would like to create/extend upon this and add their own? For instance, inline templates:

```
fn component(props map[string]string, children []some_type) some_type {
  return tpl {
    <u>foo</u>
    ${children}
  }
}
view := tpl {
  <b>${some variable}</b>
  <component/>
}
```

Implementing a DSL handler for usecases outside the web might also be useful.

Although, this is "might be"... Thought I'd put this out there and see what people think. I, for instance, would love to use a feature like this to implement a better embedding of long script segments in foreign languages. To generate a custom script, for instance, and immediately assign it to a variable or initialize a struct with it.

**Comment by MarcTCruz** [comment]
Nice feature, it may act as an attractive for developers and many business in future.

**Comment by JalonSolov** [comment]
https://discord.com/channels/592103645835821068/763496422166626336/772540283979431936

**Comment by IngwiePhoenix** [comment]
The discord link translates to:

> [20:19] Alex M: @pep DSLs also go against another V's philosophy: readable and maintainable code that's similar across companies and teams.
> With DSLs each team can write their own small language on top of V, and from my experience in the field, people will abuse it :slight_smile:
> [20:21] Alex M: But I do understand the benefits, perhaps we can figure something out.

Not everyone uses Discord, so I decided to just quote it here. :)

**Comment by smartiniOnGitHub** [comment]
Hi all, in modern JavaScript it's possible to define custom tagged templates for strings, see [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) to perform a custom elaboration of some strings; not so general but maybe could be simpler to implement and useful in some cases. In Scala there is a similar feature for strings. 
Maybe this solution could be a good compromise, at least for the beginning.

In Scala this feature is implemented and it's used in a simple way to define custom DSL, for example look [here](https://mostafa-asg.github.io/post/writing-internal-dsl-in-scala/), [here](https://stackoverflow.com/questions/49216312/what-is-dsl-in-scala), etc, with functions that accept an additional code block as argument, but could be not so simple to implement here (no idea).

Hope this helps. 
Guys what do you think ?

Bye

**Comment by JalonSolov** [comment]
This has also been done for a while in Groovy and other languages.  Doesn't make it easy to do in V, in a safe way, that doesn't lead to confusion and debugging nightmares.

Yes, it would be good to have something like this, but I would rather it be done as simply and safely as possible.
